include/networkit/algebraic/CSRGeneralMatrix.hpp

```diff
@@ -0,0 +1,1313 @@
+                          // be present
+        index mid = (left + right) / 2;
+        if (columnIdx[mid] == j) {
+            return mid;
+        } else if (columnIdx[mid] > j && mid > 0) {
+            return binarySearchColumns(left, mid - 1, j);
+        } else {
+            return binarySearchColumns(mid + 1, right, j);
+        }
+    }
+
+public:
+    /** Default constructor */
+    CSRGeneralMatrix()
+        : rowIdx(0), columnIdx(0), nonZeros(0), nRows(0), nCols(0),
+          isSorted(true), zero(0) {}
+
+    /**
+     * Constructs the CSRGeneralMatrix with size @a dimension x @a dimension.
+     * @param dimension Defines how many rows and columns this matrix has.
+     * @param zero The zero element (default = 0).
+     */
+    CSRGeneralMatrix(const count dimension, const ValueType zero = 0)
+        : rowIdx(dimension + 1), columnIdx(0), nonZeros(0), nRows(dimension),
+          nCols(dimension), isSorted(true), zero(zero) {}
+
+    /**
+     * Constructs the CSRGeneralMatrix with size @a nRows x @a nCols.
+     * @param nRows Number of rows.
+     * @param nCols Number of columns.
+     * @param zero The zero element (default = 0).
+     */
+    CSRGeneralMatrix(const count nRows, const count nCols,
+                     const ValueType zero = 0)
+        : rowIdx(nRows + 1), columnIdx(0), nonZeros(0), nRows(nRows),
+          nCols(nCols), isSorted(true), zero(zero) {}
+
+    /**
+     * Constructs the @a dimension x @a dimension Matrix from the elements at
+     * position @a positions with values @values.
+     * @param dimension Defines how many rows and columns this matrix has.
+     * @param triplets The nonzero elements.
+     * @param zero The zero element (default is 0).
+     * @param isSorted True, if the triplets are sorted per row. Default is
+     * false.
+     */
+    CSRGeneralMatrix(const count dimension,
+                     const std::vector<Triplet> &triplets,
+                     const ValueType zero = 0, bool isSorted = false)
+        : CSRGeneralMatrix(dimension, dimension, triplets, zero, isSorted) {}
+
+    /**
+     * Constructs the @a nRows x @a nCols Matrix from the elements at position
+     * @a positions with values @values.
+     * @param nRows Defines how many rows this matrix has.
+     * @param nCols Defines how many columns this matrix has.
+     * @param triplets The nonzero elements.
+     * @param zero The zero element (default is 0).
+     * @param isSorted True, if the triplets are sorted per row. Default is
+     * false.
+     */
+    CSRGeneralMatrix(const count nRows, const count nCols,
+                     const std::vector<Triplet> &triplets,
+                     const ValueType zero = 0, bool isSorted = false)
+        : nRows(nRows), nCols(nCols), isSorted(isSorted), zero(zero) {
+        count nnz = triplets.size();
+        rowIdx = std::vector<index>(nRows + 1, 0);
+        columnIdx = std::vector<index>(nnz);
+        nonZeros = std::vector<double>(nnz);
```

Comment:
Actually, looking at the code, there are still a lot of places, where `double` is used. This does not look intentional?

Reply:
My bad, it is not intentional, I'll change them to `ValueType`.

> LGTM now. I did not review the algorithm; I assume that they did not change except for changing `double` -> `ValueType` (?).

Exactly, apart from that there are no further changes.